vendor/k8s.io/system-validators/validators/types.go

```diff
@@ -112,8 +120,10 @@ type SysSpec struct {
 	OS string `json:"os,omitempty"`
 	// KernelConfig defines the spec for kernel.
 	KernelSpec KernelSpec `json:"kernelSpec,omitempty"`
-	// Cgroups is the required cgroups.
-	Cgroups []string `json:"cgroups,omitempty"`
```

Comment:
this field name change and json tag change is not backwards compatible… this should not have been done in the v1.x version. Is there a way to add the desired capabilities while remaining backwards compatible?

Also, if this serialization change ends up being required, where/how do we use this in a serialized way? Do we need to migrate data with the old serialization?

Reply:
> this field name change and json tag change is not backwards compatible… this should not have been done in the v1.x version.

oversight on my part; will restore the old field name.

> Is there a way to add the desired capabilities while remaining backwards compatible?

yes, by keeping `Cgroups (cgroups)` and adding `CgroupsV2 (cgroupsv2)`.

> Also, if this serialization change ends up being required, where/how do we use this in a serialized way? 

we can preserve the tag name for the old field.
we unmarshall in node_e2e, but do not marshal:
https://github.com/kubernetes/kubernetes/blob/edbbb6a89f9583f18051218b1adef1def1b777ae/test/e2e_node/e2e_node_suite_test.go#L343-L344

> Do we need to migrate data with the old serialization?

i don't think we do. new specs on disk would need to include the new cgroup v2 field.

Reply:
> oversight on my part; will restore the old field name.

sounds good... not just the field name, but the type (`[]string`) and meaning ("required cgroups") as well, right?

Reply:
oh, in that case this library needs a v2.0.0 release.
the "required", "optional" fields are needed at this point:
https://github.com/kubernetes/kubernetes/pull/89901/files/0c7ab806db594f22a6f5edb9c6cced3248e256d7#diff-fe0b76996791ed693a48ae1d26527c49R58-R61

Reply:
that could still be done compatibly with parallel fields (e.g. `Cgroups`, `OptionalCgroups`, `V2CGroups`, `OptionalV2Cgroups`)

Reply:
in terms of migration, my take is that we should not be providing one as node_e2e is the only known consumer of the serialized spec.

i'm hoping that the cgroupv2 contributors will make the required changes in 1.19.

Reply:
> that could still be done compatibly with parallel fields (e.g. Cgroups, OptionalCgroups, V2Cgroups, OptionalV2Cgroups)

ok, this can work.